Added snapcraft to build and workflow for release

gabrielcocenza opened the pull request:
- hooks and local folders necessary to build the snap were copied from the upstream, but the maintenance will be at this repo.
- workflows include workflow_dispatch to be able to manually trigger the release to latest/edge.

gabrielcocenza committed:
Added snapcraft to build and workflow for release

- hooks and local folders necessary to build the snap were copied
  from the upstream, but the maintenance will be at this repo.
- workflows include workflow_dispatch to be able to manually
  trigger the release to latest/edge.

samuelallan72 commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,30 @@
+          SNAPCRAFT_STORE_CREDENTIALS: ${{ secrets.STORE_LOGIN }}
+        with:
+          snap: ${{ steps.build.outputs.snap }}
+          release: latest/edge
```
Please add the trailing newline. :)

samuelallan72 commented on snap/hooks/configure:
```diff
@@ -0,0 +1,67 @@
+#!/bin/sh -e
```
Since these are copied from upstream without change, did you want to retain your script for syncing them?  That might help us keep up to date in future?

gabrielcocenza replied:
I thought about that and I think it will be very rare that someone outside of Canonical will change the scripts to build the charm. IMO I think this repo should be the source of "truth" to anything related to the snap package instead of doing on the upstream repo.

Regarding the name, it's unusual to have the `snap` at the name so I'm in favor of keeping this way.

cc: @aieri

gabrielcocenza replied:
On the stand up meeting Andrea agrees with your proposal, so I'm going to change the name and sync the script from the upstream

samuelallan72 reviewed (changes requested):
Thanks for this!  It looks pretty good; I just had a couple of comments.

Also this repo should be named `charmed-openstack-exporter-snap` or `snap-charmed-openstack-exporter` to avoid confusion.  This repo only provides a snap; it's not an exporter project in itself.

gabrielcocenza committed:
- add trailing newline

gabrielcocenza committed:
- added sync script to update necessary files to build the snap
- added makefile to be able to easily build locally

samuelallan72 commented on Makefile:
```diff
@@ -0,0 +1,12 @@
+help:
+	@echo "This project supports the following targets"
+	@echo ""
+	@echo " make help - show this text"
+	@echo " make build - build the snap"
+	@echo ""
+
+build:
+	@echo update snap hooks and local
+	@bash -c ./sync_upstream.sh
+	@echo "Building the snap"
+	@snapcraft --use-lxd
```
I'm in favour of removing this Makefile.  I think the sync from upstream script is for convenience only if we ever want to resync from upstream; we don't want to run it every time we build the snap.

Better to simply document:

- `snapcraft` to build the snap
- `./sync_upstream.sh` to update the files from upstream if needed at some point

samuelallan72 reviewed (changes requested):
Looks good to me, thanks!  I'm not sure about the makefile though.

chanchiwai-ray commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,49 @@
+    build-snaps:
+      - go
+    # override pull to read VERSION because working dir is root of source repo
+    override-pull: |
+      craftctl default
+      craftctl set version="$(cat VERSION)"
```
These need to be removed because the `VERSION` file is only available in the upstream project. Or it can be improved e.g.

Edit: this might need `source-tags` see later comment
```
    override-pull: |
      craftctl default
      TAG=$(git describe --tags --abbrev=10)
      craftctl set version="${TAG#v}"
```
or

(this need annotated tags)

Edit: this might not work because this might get the tag from this repository.
```
version: git
```

gabrielcocenza replied:
Nice catch. It was working to cat the `VERSION` from the upstream, however it's more accurate to get from the git itself. E.g: on upstream is using `1.6.0`, but the current version is `1.7.x`. Probably the maintainer forgot to update the `VERSION` file. I think we should change also in upstream to be more realistic in the version, otherwise is very easy to forget when building the snap and releasing it.

samuelallan72 replied:
Agreed, if we make improvements to the snapcraft file, it would be nice to contribute upstream.

@chanchiwai-ray is referring to this upstream issue: https://github.com/openstack-exporter/openstack-exporter/issues/357 - this suggestion will fix that.  :)

chanchiwai-ray commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,49 @@
+  openstack-exporter:
+    plugin: go
+    source: https://github.com/openstack-exporter/openstack-exporter.git
+    source-type: git
```
I think it's better add `source-tags`, but it depends on how do we want to do the release (always latest change or use stable one)

gabrielcocenza replied:
In the current case we need to use the latest change because the last stable one doesn't include the cache introduction

chanchiwai-ray replied:
got it

gabrielcocenza committed:
- remove makefile
- change the way to set the version of the snap by git tag of the
  latest commit.
